File: templates/python/anr/anr.py
"""
ANR — Agent Network Record
──────────────────────────────────────────────────────────────────────────────
Python reference implementation.

Wire format  : RLP( [sig, seq, k₁, v₁, k₂, v₂, …] )
Signed over  : keccak256( RLP( [b"anr-v1", seq_bytes, k₁, v₁, …] ) )
Text form    : "anr:" + base64url(wire, no padding)
Max size     : 512 bytes

Standard ANR keys
─────────────────
  Network  : id  secp256k1  ip  ip6  tcp  udp
  Agent    : a.id  a.name  a.ver  a.caps  a.tags
             a.proto  a.port  a.tls  a.meta  a.owner  a.chain

Install deps
────────────
  pip install eth-hash[pycryptodome] eth-rlp coincurve
"""

from __future__ import annotations

import struct
import base64
from dataclasses import dataclass, field
from typing import Optional

# ── optional crypto deps (install separately) ─────────────────────────────────
try:
    from eth_hash.auto import keccak
    import coincurve
    _CRYPTO_OK = True
except ImportError:
    _CRYPTO_OK = False
    import warnings
    warnings.warn("coincurve / eth-hash not installed — sign/verify disabled")

# ── RLP (minimal self-contained implementation) ───────────────────────────────

def _rlp_length_prefix(length: int, offset: int) -> bytes:
    if length < 56:
        return bytes([offset + length])
    enc = length.to_bytes((length.bit_length() + 7) // 8, 'big')
    return bytes([offset + 55 + len(enc)]) + enc

def rlp_encode(item) -> bytes:
    if isinstance(item, (bytes, bytearray)):
        b = bytes(item)
        if len(b) == 1 and b[0] < 0x80:
            return b
        return _rlp_length_prefix(len(b), 0x80) + b
    if isinstance(item, list):
        payload = b''.join(rlp_encode(i) for i in item)
        return _rlp_length_prefix(len(payload), 0xC0) + payload
    raise TypeError(f"RLP cannot encode {type(item)}")

def _decode_length(data: bytes, offset: int):
    prefix = data[offset]
    if prefix < 0x80:
        return offset + 1, 1, 'str'
    if prefix <= 0xB7:
        length = prefix - 0x80
        return offset + 1, length, 'str'
    if prefix <= 0xBF:
        ll = prefix - 0xB7
        length = int.from_bytes(data[offset + 1 : offset + 1 + ll], 'big')
        return offset + 1 + ll, length, 'str'
    if prefix <= 0xF7:
        length = prefix - 0xC0
        return offset + 1, length, 'list'
    ll = prefix - 0xF7
    length = int.from_bytes(data[offset + 1 : offset + 1 + ll], 'big')
    return offset + 1 + ll, length, 'list'

def rlp_decode(data: bytes):
    def _decode(data, offset):
        start, length, kind = _decode_length(data, offset)
        if kind == 'str':
            return data[start : start + length], start + length
        items, pos = [], start
        while pos < start + length:
            item, pos = _decode(data, pos)
            items.append(item)
        return items, start + length
    result, _ = _decode(data, 0)
    return result

# ── encoding helpers ──────────────────────────────────────────────────────────

ANR_PREFIX    = "anr:"
ANR_ID_SCHEME = "amp-v1"
ANR_MAX_BYTES = 512
SIGN_DOMAIN   = b"anr-v1"


def _enc_uint16(n: int) -> bytes:
    return struct.pack(">H", n)

def _dec_uint16(b: bytes) -> int:
    return struct.unpack(">H", b)[0]

def _enc_uint64(n: int) -> bytes:
    return struct.pack(">Q", n)

def _dec_uint64(b: bytes) -> int:
    return struct.unpack(">Q", b)[0]

def _sorted_kv(kv: dict[str, bytes]) -> list[tuple[str, bytes]]:
    return sorted(kv.items(), key=lambda x: x[0])

# ── ANR dataclass ─────────────────────────────────────────────────────────────

@dataclass
class ANR:
    seq:       int
    kv:        dict[str, bytes] = field(default_factory=dict)
    signature: bytes = b'\x00' * 64   # 64-byte r‖s

    def encode(self) -> bytes:
        """Encode to binary RLP wire format."""
        pairs = []
        for k, v in _sorted_kv(self.kv):
            pairs.append(k.encode())
            pairs.append(v)
        wire = rlp_encode([self.signature, _enc_uint64(self.seq)] + pairs)
        if len(wire) > ANR_MAX_BYTES:
            raise ValueError(f"ANR exceeds max size: {len(wire)} > {ANR_MAX_BYTES}")
        return wire

    def encode_text(self) -> str:
        """Encode to canonical 'anr:<base64url>' text form."""
        wire = self.encode()
        b64  = base64.urlsafe_b64encode(wire).rstrip(b'=').decode()
        return ANR_PREFIX + b64

    @staticmethod
    def decode(wire: bytes) -> "ANR":
        """Decode from RLP bytes."""
        if len(wire) > ANR_MAX_BYTES:
            raise ValueError(f"ANR too large: {len(wire)}")
        lst = rlp_decode(wire)
        if not isinstance(lst, list) or len(lst) < 2:
            raise ValueError("Invalid ANR structure")
        sig_bytes, seq_bytes, *rest = lst
        kv: dict[str, bytes] = {}
        for i in range(0, len(rest), 2):
            kv[rest[i].decode()] = rest[i + 1]
        return ANR(seq=_dec_uint64(seq_bytes), kv=kv, signature=sig_bytes)

    @staticmethod
    def decode_text(text: str) -> "ANR":
        """Decode from 'anr:<base64url>' text form."""
        if not text.startswith(ANR_PREFIX):
            raise ValueError(f"ANR text must start with '{ANR_PREFIX}'")
        # Re-add padding
        b64_part = text[len(ANR_PREFIX):]
        pad      = 4 - len(b64_part) % 4
        if pad != 4:
            b64_part += '=' * pad
        wire = base64.urlsafe_b64decode(b64_part)
        return ANR.decode(wire)

    def verify(self) -> bool:
        """Verify the record's signature against its stored public key."""
        if not _CRYPTO_OK:
            raise RuntimeError("coincurve / eth-hash not installed")
        pubkey_bytes = self.kv.get('secp256k1')
        if not pubkey_bytes:
            return False
        try:
            content = _content_rlp(self.seq, self.kv)
            digest  = keccak(content)
            pub     = coincurve.PublicKey(pubkey_bytes)
            return pub.verify(self.signature, digest, hasher=None)
        except Exception:
            return False

    def parsed(self) -> "ParsedANR":
        """Decode well-known fields into a typed object."""
        return ParsedANR.from_anr(self)


def _content_rlp(seq: int, kv: dict[str, bytes]) -> bytes:
    pairs = []
    for k, v in _sorted_kv(kv):
        pairs.append(k.encode())
        pairs.append(v)
    return rlp_encode([SIGN_DOMAIN, _enc_uint64(seq)] + pairs)


# ── signing ───────────────────────────────────────────────────────────────────

def sign_anr(private_key_bytes: bytes, seq: int, kv: dict[str, bytes]) -> ANR:
    """Create a signed ANR from a raw secp256k1 private key (32 bytes)."""
    if not _CRYPTO_OK:
        raise RuntimeError("coincurve not installed — pip install coincurve")
    priv    = coincurve.PrivateKey(private_key_bytes)
    pub     = priv.public_key.format(compressed=True)          # 33 bytes
    kv['id']        = ANR_ID_SCHEME.encode()
    kv['secp256k1'] = pub
    content  = _content_rlp(seq, kv)
    digest   = keccak(content)
    sig      = priv.sign(digest, hasher=None)[:64]             # drop recovery byte
    return ANR(seq=seq, kv=kv, signature=sig)


# ── builder ───────────────────────────────────────────────────────────────────

class AnrBuilder:
    """Fluent builder for constructing ANRs without raw byte manipulation."""

    def __init__(self):
        self._seq: int = 0
        self._kv: dict[str, bytes] = {}

    def seq(self, n: int)                -> "AnrBuilder": self._seq = n; return self
    def agent_id(self, v: str)           -> "AnrBuilder": self._kv['a.id']    = v.encode(); return self
    def name(self, v: str)               -> "AnrBuilder": self._kv['a.name']  = v.encode(); return self
    def version(self, v: str)            -> "AnrBuilder": self._kv['a.ver']   = v.encode(); return self
    def capabilities(self, caps: list[str]) -> "AnrBuilder":
        self._kv['a.caps'] = rlp_encode([c.encode() for c in caps]); return self
    def tags(self, tags: list[str])      -> "AnrBuilder":
        self._kv['a.tags'] = rlp_encode([t.encode() for t in tags]); return self
    def proto(self, v: str)              -> "AnrBuilder": self._kv['a.proto'] = v.encode(); return self
    def agent_port(self, port: int)      -> "AnrBuilder": self._kv['a.port']  = _enc_uint16(port); return self
    def tls(self, enabled: bool)         -> "AnrBuilder": self._kv['a.tls']   = bytes([1 if enabled else 0]); return self
    def meta_uri(self, uri: str)         -> "AnrBuilder": self._kv['a.meta']  = uri.encode(); return self
    def owner(self, addr_bytes: bytes)   -> "AnrBuilder": self._kv['a.owner'] = addr_bytes; return self
    def chain_id(self, cid: int)         -> "AnrBuilder": self._kv['a.chain'] = _enc_uint64(cid); return self
    def ipv4(self, b: bytes)             -> "AnrBuilder": self._kv['ip']  = b; return self
    def ipv6(self, b: bytes)             -> "AnrBuilder": self._kv['ip6'] = b; return self
    def tcp_port(self, port: int)        -> "AnrBuilder": self._kv['tcp'] = _enc_uint16(port); return self
    def udp_port(self, port: int)        -> "AnrBuilder": self._kv['udp'] = _enc_uint16(port); return self

    def sign(self, private_key: bytes) -> ANR:
        return sign_anr(private_key, self._seq, dict(self._kv))


# ── parsed view ───────────────────────────────────────────────────────────────

@dataclass
class ParsedANR:
    seq:          int
    id_scheme:    Optional[str]
    pubkey:       Optional[bytes]
    agent_id:     Optional[str]
    name:         Optional[str]
    version:      Optional[str]
    capabilities: list[str]
    tags:         list[str]
    proto:        Optional[str]
    agent_port:   Optional[int]
    tls:          bool
    meta_uri:     Optional[str]
    owner:        Optional[bytes]
    chain_id:     Optional[int]
    ip:           Optional[bytes]
    ip6:          Optional[bytes]
    tcp_port:     Optional[int]
    udp_port:     Optional[int]

    @staticmethod
    def from_anr(r: ANR) -> "ParsedANR":
        def s(k):
            v = r.kv.get(k)
            return v.decode() if v else None

        def decode_list(k):
            raw = r.kv.get(k)
            if not raw:
                return []
            try:
                items = rlp_decode(raw)
                return [i.decode() for i in items]
            except Exception:
                return []

        return ParsedANR(
            seq          = r.seq,
            id_scheme    = s('id'),
            pubkey       = r.kv.get('secp256k1'),
            agent_id     = s('a.id'),
            name         = s('a.name'),
            version      = s('a.ver'),
            capabilities = decode_list('a.caps'),
            tags         = decode_list('a.tags'),
            proto        = s('a.proto'),
            agent_port   = _dec_uint16(r.kv['a.port']) if 'a.port' in r.kv else None,
            tls          = r.kv.get('a.tls', b'\x00')[0] == 1,
            meta_uri     = s('a.meta'),
            owner        = r.kv.get('a.owner'),
            chain_id     = _dec_uint64(r.kv['a.chain']) if 'a.chain' in r.kv else None,
            ip           = r.kv.get('ip'),
            ip6          = r.kv.get('ip6'),
            tcp_port     = _dec_uint16(r.kv['tcp']) if 'tcp' in r.kv else None,
            udp_port     = _dec_uint16(r.kv['udp']) if 'udp' in r.kv else None,
        )


# ── smoke test ────────────────────────────────────────────────────────────────

if __name__ == '__main__':
    import os, json

    priv = os.urandom(32)   # random key — use a fixed key in production

    record = (
        AnrBuilder()
        .seq(1)
        .agent_id('sentrix://agent/0xABC123')
        .name('WeatherAgent')
        .version('1.0.0')
        .capabilities(['getWeather', 'getForecast'])
        .tags(['weather', 'data'])
        .proto('http')
        .agent_port(6174)
        .tls(False)
        .meta_uri('ipfs://QmWeatherMeta')
        .ipv4(bytes([127, 0, 0, 1]))
        .tcp_port(9000)
        .sign(priv)
    )

    text    = record.encode_text()
    decoded = ANR.decode_text(text)
    parsed  = decoded.parsed()

    print("=== ANR text ===")
    print(text)
    print("\n=== Parsed ===")
    print(json.dumps({
        'seq':          parsed.seq,
        'agent_id':     parsed.agent_id,
        'name':         parsed.name,
        'version':      parsed.version,
        'capabilities': parsed.capabilities,
        'tags':         parsed.tags,
        'proto':        parsed.proto,
        'agent_port':   parsed.agent_port,
        'tls':          parsed.tls,
    }, indent=2))

    if _CRYPTO_OK:
        valid = decoded.verify()
        print(f"\n✔ Signature valid: {valid}")
        assert valid

    print("\n✔ Round-trip OK")
    assert parsed.name == 'WeatherAgent'
    assert 'getWeather' in parsed.capabilities

File: templates/typescript/anr/anr.ts
/**
 * ANR — Agent Network Record
 * ───────────────────────────────────────────────────────────────────────────
 * Inspired by EIP-778 (Ethereum Node Records).
 *
 * Wire format  : RLP( [sig, seq, k₁, v₁, k₂, v₂, …] )
 * Signed over  : keccak256( RLP( ["anr-v1", seq, k₁, v₁, …] ) )
 * Text form    : "anr:" + base64url(wire bytes, no padding)
 * Max size     : 512 bytes
 * Key ordering : lexicographic, unique, no duplicates
 *
 * Standard ANR keys
 * ─────────────────
 *  Network (inherited from ENR):
 *    id         identity scheme name ("amp-v1")
 *    secp256k1  compressed public key (33 bytes)
 *    ip         IPv4 address (4 bytes)
 *    ip6        IPv6 address (16 bytes)
 *    tcp        TCP port (uint16 big-endian)
 *    udp        UDP port (uint16 big-endian)
 *
 *  Agent-specific  (prefix "a."):
 *    a.id       agent identifier string  e.g. "borgkit://agent/0xABC"
 *    a.name     human-readable name
 *    a.ver      semver string  e.g. "1.2.3"
 *    a.caps     RLP list of capability name strings
 *    a.tags     RLP list of tag strings
 *    a.proto    transport hint  "http" | "ws" | "grpc" | "tcp"
 *    a.port     agent API port (uint16 big-endian)
 *    a.tls      TLS flag  0x00 | 0x01
 *    a.meta     IPFS / Arweave metadata URI
 *    a.owner    owner wallet address (20 bytes)
 *    a.chain    EVM chain ID (uint64 big-endian)
 */

import { keccak256 } from 'ethereum-cryptography/keccak';
import { secp256k1 } from 'ethereum-cryptography/secp256k1';
import { RLP } from '@ethereumjs/rlp';

// ── constants ────────────────────────────────────────────────────────────────

export const ANR_PREFIX   = 'anr:';
export const ANR_ID_SCHEME = 'amp-v1';
export const ANR_MAX_BYTES = 512;
const SIGN_DOMAIN = new TextEncoder().encode('anr-v1');

// ── types ────────────────────────────────────────────────────────────────────

/** All known ANR keys — callers may also use arbitrary string keys. */
export type AnrKey =
  | 'id' | 'secp256k1' | 'ip' | 'ip6' | 'tcp' | 'udp'
  | 'a.id' | 'a.name' | 'a.ver' | 'a.caps' | 'a.tags'
  | 'a.proto' | 'a.port' | 'a.tls' | 'a.meta' | 'a.owner' | 'a.chain'
  | string;

export type AnrKV = Map<string, Uint8Array>;

export interface ANR {
  /** Sequence number — increment on every update */
  seq: bigint;
  /** Key-value pairs (sorted lexicographically before encoding) */
  kv: AnrKV;
  /** 64-byte secp256k1 signature (r‖s) */
  signature: Uint8Array;
}

// ── helpers ──────────────────────────────────────────────────────────────────

function encodeUint16(n: number): Uint8Array {
  const buf = new Uint8Array(2);
  new DataView(buf.buffer).setUint16(0, n, false); // big-endian
  return buf;
}

function encodeUint64(n: bigint): Uint8Array {
  const buf = new Uint8Array(8);
  new DataView(buf.buffer).setBigUint64(0, n, false);
  return buf;
}

function decodeUint16(b: Uint8Array): number {
  return new DataView(b.buffer, b.byteOffset).getUint16(0, false);
}

function decodeUint64(b: Uint8Array): bigint {
  return new DataView(b.buffer, b.byteOffset).getBigUint64(0, false);
}

function sortedKV(kv: AnrKV): Array<[string, Uint8Array]> {
  return [...kv.entries()].sort(([a], [b]) => a < b ? -1 : a > b ? 1 : 0);
}

// ── content (what gets signed) ────────────────────────────────────────────────

function contentRlp(seq: bigint, kv: AnrKV): Uint8Array {
  const pairs = sortedKV(kv).flatMap(([k, v]) => [
    new TextEncoder().encode(k),
    v,
  ]);
  return RLP.encode([SIGN_DOMAIN, encodeUint64(seq), ...pairs]);
}

// ── sign / verify ─────────────────────────────────────────────────────────────

/**
 * Create a signed ANR from a private key.
 * @param privateKey  32-byte secp256k1 private key
 * @param seq         Sequence number
 * @param kv          Key-value map (automatically receives `id` and `secp256k1`)
 */
export function createANR(
  privateKey: Uint8Array,
  seq: bigint,
  kv: Map<string, Uint8Array>
): ANR {
  const pubkey = secp256k1.getPublicKey(privateKey, true); // 33-byte compressed

  kv.set('id',        new TextEncoder().encode(ANR_ID_SCHEME));
  kv.set('secp256k1', pubkey);

  const hash = keccak256(contentRlp(seq, kv));
  const sig  = secp256k1.sign(hash, privateKey);
  const signature = sig.toCompactRawBytes(); // 64 bytes (r‖s)

  return { seq, kv, signature };
}

/**
 * Verify the signature of a decoded ANR.
 * Returns true if valid.
 */
export function verifyANR(record: ANR): boolean {
  const pubkeyBytes = record.kv.get('secp256k1');
  if (!pubkeyBytes) return false;
  try {
    const hash = keccak256(contentRlp(record.seq, record.kv));
    const sig  = secp256k1.Signature.fromCompact(record.signature);
    return secp256k1.verify(sig, hash, pubkeyBytes);
  } catch {
    return false;
  }
}

// ── encode ────────────────────────────────────────────────────────────────────

/**
 * Encode an ANR to its binary RLP wire format.
 */
export function encodeANR(record: ANR): Uint8Array {
  const pairs = sortedKV(record.kv).flatMap(([k, v]) => [
    new TextEncoder().encode(k),
    v,
  ]);
  const wire = RLP.encode([record.signature, encodeUint64(record.seq), ...pairs]);

  if (wire.length > ANR_MAX_BYTES) {
    throw new Error(`ANR exceeds max size: ${wire.length} > ${ANR_MAX_BYTES} bytes`);
  }
  return wire;
}

/**
 * Encode an ANR to its canonical text form: "anr:<base64url>"
 */
export function encodeANRText(record: ANR): string {
  const wire    = encodeANR(record);
  const b64     = Buffer.from(wire).toString('base64url');
  return ANR_PREFIX + b64;
}

// ── decode ────────────────────────────────────────────────────────────────────

/**
 * Decode ANR from raw binary RLP bytes.
 * Throws if structure is invalid.
 */
export function decodeANR(wire: Uint8Array): ANR {
  if (wire.length > ANR_MAX_BYTES) {
    throw new Error(`ANR exceeds max size: ${wire.length} bytes`);
  }

  const list = RLP.decode(wire) as Uint8Array[];
  if (!Array.isArray(list) || list.length < 2 || list.length % 2 !== 0) {
    throw new Error('Invalid ANR RLP structure');
  }

  const [sigBytes, seqBytes, ...rest] = list;
  if (rest.length % 2 !== 0) throw new Error('ANR key-value pairs must be even');

  const kv: AnrKV = new Map();
  for (let i = 0; i < rest.length; i += 2) {
    const key = new TextDecoder().decode(rest[i]);
    kv.set(key, rest[i + 1]);
  }

  return {
    seq:       decodeUint64(seqBytes),
    kv,
    signature: sigBytes,
  };
}

/**
 * Decode ANR from its "anr:<base64url>" text form.
 */
export function decodeANRText(text: string): ANR {
  if (!text.startsWith(ANR_PREFIX)) {
    throw new Error(`ANR text must start with "${ANR_PREFIX}"`);
  }
  const wire = Buffer.from(text.slice(ANR_PREFIX.length), 'base64url');
  return decodeANR(wire);
}

// ── high-level builder ────────────────────────────────────────────────────────

/** Typed builder — avoids raw byte manipulation for common fields. */
export class AnrBuilder {
  private kv: AnrKV = new Map();
  private seq: bigint = 0n;

  setSeq(seq: bigint)             { this.seq = seq; return this; }

  // Agent fields
  setAgentId(id: string)          { this.kv.set('a.id',    enc(id));     return this; }
  setName(name: string)           { this.kv.set('a.name',  enc(name));   return this; }
  setVersion(ver: string)         { this.kv.set('a.ver',   enc(ver));    return this; }
  setCapabilities(caps: string[]) { this.kv.set('a.caps',  RLP.encode(caps.map(enc))); return this; }
  setTags(tags: string[])         { this.kv.set('a.tags',  RLP.encode(tags.map(enc))); return this; }
  setProto(proto: string)         { this.kv.set('a.proto', enc(proto));  return this; }
  setAgentPort(port: number)      { this.kv.set('a.port',  encodeUint16(port)); return this; }
  setTls(tls: boolean)            { this.kv.set('a.tls',   new Uint8Array([tls ? 1 : 0])); return this; }
  setMetaUri(uri: string)         { this.kv.set('a.meta',  enc(uri));    return this; }
  setOwner(addr: Uint8Array)      { this.kv.set('a.owner', addr);        return this; }
  setChainId(id: bigint)          { this.kv.set('a.chain', encodeUint64(id)); return this; }

  // Network fields
  setIpv4(bytes: Uint8Array)      { this.kv.set('ip',  bytes); return this; }
  setIpv6(bytes: Uint8Array)      { this.kv.set('ip6', bytes); return this; }
  setTcpPort(port: number)        { this.kv.set('tcp', encodeUint16(port)); return this; }
  setUdpPort(port: number)        { this.kv.set('udp', encodeUint16(port)); return this; }

  sign(privateKey: Uint8Array): ANR {
    return createANR(privateKey, this.seq, this.kv);
  }
}

// ── decoder helpers ───────────────────────────────────────────────────────────

/** Decode the well-known fields from a verified ANR into a plain object. */
export function parseANRFields(record: ANR): ParsedANR {
  const { kv } = record;
  const get = (k: string) => kv.get(k);
  const str = (k: string) => { const v = get(k); return v ? new TextDecoder().decode(v) : undefined; };

  return {
    seq:          record.seq,
    idScheme:     str('id'),
    pubkey:       get('secp256k1'),
    agentId:      str('a.id'),
    name:         str('a.name'),
    version:      str('a.ver'),
    capabilities: decodeCaps(get('a.caps')),
    tags:         decodeCaps(get('a.tags')),
    proto:        str('a.proto'),
    agentPort:    get('a.port')  ? decodeUint16(get('a.port')!)  : undefined,
    tls:          get('a.tls')   ? get('a.tls')![0] === 1         : false,
    metaUri:      str('a.meta'),
    owner:        get('a.owner'),
    chainId:      get('a.chain') ? decodeUint64(get('a.chain')!) : undefined,
    ip:           get('ip'),
    ip6:          get('ip6'),
    tcpPort:      get('tcp')     ? decodeUint16(get('tcp')!)     : undefined,
    udpPort:      get('udp')     ? decodeUint16(get('udp')!)     : undefined,
  };
}

export interface ParsedANR {
  seq:          bigint;
  idScheme?:    string;
  pubkey?:      Uint8Array;
  agentId?:     string;
  name?:        string;
  version?:     string;
  capabilities: string[];
  tags:         string[];
  proto?:       string;
  agentPort?:   number;
  tls:          boolean;
  metaUri?:     string;
  owner?:       Uint8Array;
  chainId?:     bigint;
  ip?:          Uint8Array;
  ip6?:         Uint8Array;
  tcpPort?:     number;
  udpPort?:     number;
}

// ── internal ──────────────────────────────────────────────────────────────────

function enc(s: string): Uint8Array { return new TextEncoder().encode(s); }

function decodeCaps(raw: Uint8Array | undefined): string[] {
  if (!raw) return [];
  try {
    const list = RLP.decode(raw) as Uint8Array[];
    return list.map(b => new TextDecoder().decode(b));
  } catch {
    return [];
  }
}
